### PR TITLE
feat(APIv2): RHINENG-2198 introduce the derived_attribute serializer

### PR DESCRIPTION
**Pull request by skateman**
This change introduces a `fields` method on the serializer(s) that allow the user to determine what fields are required from a given model to construct the JSON serialized result returned by the controller. Furthermore, a new `derived_attribute` is introduced to the serializer's DSL for declaring fields that depend on other models. The logic deciding what fields should be shown has been implemented using Panko's [`.filter_for`](https://panko.dev/docs/attributes#filters-for) method. Finally, the rendering is adjusted to only select & alias fields that are really required for serialization. 

The new `derived_attribute` should be used to declare a field of an ActiveRecord object that is conditionally constructed by data from the parent (joined) objects if they are available. A good example of this is the `Rule#remediation_issue_id` which should be only available when the `Rule` objects are joined with both `Profile` and `SecurityGuide` objects, i.e. under `/security_guides/:id/profiles/:id/rules`.

```
derived_attribute :os_major_version, :has_remediation, profiles: [:ref_id], security_guide: [:ref_id]
```

The first argument is the name of the method, next is an optional array of dependent fields on the current model, then a hash of dependent `associations: [fields]`. The generated attribute is automatically delegated to a method with the same name defined in the model, so in the case of the example a `V2::Rule#remediation_issue_id` methods needs to exist.

Based on what is declared in the serializer, we can construct a list of fields to be queried, therefore, this also improves DB performance. The dependencies make sure that the field is only rendered in cases when the required auto-joined dependencies are available. 

## Secure Coding Practices Checklist GitHub Link
- https://github.com/RedHatInsights/secure-coding-checklist

## Secure Coding Checklist
- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [x] General Coding Practices

### CODE REVIEW

**Comment by skateman** [comment]
@vkrizan I am not fully done with the feature (the implementation of the `remediation_issue_id` is still missing), but I am trying to stabilize the API of the DSL, so I'd like your input on it.

File: app/serializers/v2/rule_serializer.rb [hunk 1]
```diff
@@ -4,5 +4,7 @@ module V2
   # JSON serialization for an OpenSCAP Rule
   class RuleSerializer < V2::ApplicationSerializer
     attributes :ref_id, :title, :rationale, :description, :severity, :precedence
+
+    complex_attribute :remediation_issue_id, profiles: [:ref_id], security_guide: [:ref_id]
```
Comment (vkrizan) [inline app/serializers/v2/rule_serializer.rb:8]: What about naming it `dependent_attribute`?

**Review by vkrizan — COMMENTED** [review]
I'd need a bit more time to go through it and understand the need to do it this way.

My concern is that we're creating something very custom, which adds complexity and future maintenance costs. If there would be a different way of resolving the problem that would be simpler, I'd choose that.

For example I'd like to know why this pushes the problem to the sterilizer. My other solution was to implement such computed fields (like `remediation_issue_id`) directly on the model. I think that model can query the scope to find out whether data that it requires to compute it are available. This seems a bit more simpler to me, as it does not crosses the boundary of a model (which is an ActiveRecord). However, it would need to be done for each computed field that would need to have performance problem improved.

File: app/serializers/v2/rule_serializer.rb (hunk 1 above)
Comment (skateman) [reply app/serializers/v2/rule_serializer.rb:8]: Longer 😁

**Comment by skateman** [comment]
The only place we are concerned with performance is during serialization. The idea is that you can (and should be able to) call `remediation_issue_id` on the model itself without the dependencies available and then the model would do the querying for you. This is standard behavior in rails, what we are doing is making this fast when we ~~sterilize~~ serialize stuff.

**Comment by vkrizan** [comment]
> The only place we are concerned with performance is during serialization.

Sure. The same is true to say about API responses. But I hope that this is not the reason why to fix it on the serializer.

> The idea is that you can (and should be able to) call remediation_issue_id on the model itself without the dependencies available and then the model would do the querying for you. This is standard behavior in rails, ...

The idea is that the model already knows about querying, hence it can know whether or not it has the required data, and stop itself from (auto)querying it to avoid the N+1 query problem.

**Comment by skateman** [comment]
The model is not aware of this and I am not sure if it is a good idea to put this there, the query interface is separated from the models on purpose and its intended place to be called is in controllers. From this perspective the serializer is a helper service in the controller, where it makes more sense to hold such logic.

On the other hand, even if the model deals with this logic, the serializer still has to be aware of the absence of this field. Setting it to `nil` on the model and hiding it is not an option as in some cases we do want to show empty field values and it is not a good idea to tangle these two. Even though the serializer has some kind of special value that we can set if the value should be hidden, I don't like the idea of tangling the serializer code into the model as we might request this data in other places than the serializer.

File: app/models/application_record.rb [hunk 2]
```diff
@@ -27,4 +27,10 @@ def self.taggable?
   def self.count_by
     primary_key.to_sym
   end
+
+  private
+
+  def attributes_available?(*attrs)
```
Comment (romanblanco) [inline app/models/application_record.rb:33]: @skateman this method is not used

File: app/serializers/v2/rule_serializer.rb (hunk 1 above)
Comment (romanblanco) [reply app/serializers/v2/rule_serializer.rb]: I'd also lean towards `dependent_attribute`. It better describes the use.

Comment (skateman) [reply app/serializers/v2/rule_serializer.rb]: What about `derived_`?

Comment (romanblanco) [reply app/serializers/v2/rule_serializer.rb]: `derived_` works for me :+1:

Comment (vkrizan) [reply app/serializers/v2/rule_serializer.rb]: :+1:

**Pull request merged**